src/contracts/Vault.sol

```diff
@@ -0,0 +1,103 @@
+
+        _id = ++depositsCount;
+        lockedMetric[_id].withdrawer = _withdrawer;
+        lockedMetric[_id].amount = _amount;
+        lockedMetric[_id].unlockTimestamp = _unlockTimestamp;
+        lockedMetric[_id].withdrawn = false;
+        lockedMetric[_id].deposited = true;
+        lockedMetric[_id].published = false;
+        lockedMetric[_id].disqualified = false;
+
+        depositsByMetricAddress[address(_metric)].push(_id);
+        depositsByWithdrawers[_withdrawer].push(_id);
+        return _id;
+    }
+
+    function withdrawMetric(uint256 _id) external {
+        require(msg.sender == lockedMetric[_id].withdrawer, "You are not the withdrawer!");
+        require(lockedMetric[_id].deposited, "No Metric Deposited!");
+        require(lockedMetric[_id].published, "Question has not been published");
+        require(!lockedMetric[_id].withdrawn, "Metric already withdrawn!");
+
+        lockedMetric[_id].withdrawn = true;
+
+        walletMetricBalance[address(lockedMetric[_id].metric)][msg.sender] = walletMetricBalance[address(lockedMetric[_id].metric)][msg.sender].sub(
+            lockedMetric[_id].amount
+        );
+
+        emit Withdraw(msg.sender, lockedMetric[_id].amount);
+        lockedMetric[_id].metric.safeTransfer(msg.sender, lockedMetric[_id].amount);
+    }
+
+    //TODO: Double check math
+    function slashMetric(uint256 _id) external onlyOwner {
+        require(lockedMetric[_id].slashed = false);
+        walletMetricBalance[address(lockedMetric[_id].metric)][msg.sender] = walletMetricBalance[address(lockedMetric[_id].metric)][msg.sender].sub(
+            lockedMetric[_id].amount.div(2)
+        );
+
+        lockedMetric[_id].slashed = true;
+
+        emit Withdraw(msg.sender, lockedMetric[_id].amount);
+        //TODO: send to treasury
+        lockedMetric[_id].metric.safeTransfer(address(this), lockedMetric[_id].amount.div(2));
+        lockedMetric[_id].metric.safeTransfer(msg.sender, lockedMetric[_id].amount.div(2));
+    }
+
+    function getDepositsByMetricAddress(address _id) external view returns (uint256[] memory) {
+        return depositsByMetricAddress[_id];
+    }
+
+    function getDepositsByWithdrawer(address _metric, address _withdrawer) external view returns (uint256) {
+        return walletMetricBalance[_metric][_withdrawer];
+    }
+
+    function getVaultsByWithdrawer(address _withdrawer) external view returns (uint256[] memory) {
+        return depositsByWithdrawers[_withdrawer];
+    }
+
+    function getVaultById(uint256 _id) external view returns (Items memory) {
+        return lockedMetric[_id];
+    }
+
+    function getMetricTotalLockedBalance(address _metric) external view returns (uint256) {
+        return IERC20(_metric).balanceOf(address(this));
+    }
+
+    event Withdraw(address withdrawer, uint256 amount);
+
+    struct Items {
+        address withdrawer;
+        uint256 amount;
+        uint256 unlockTimestamp;
+        bool withdrawn;
```

Comment:
should this be an enum?  Are these states mutually exclusive?  it might be a pain in the ass to micromanage all of em together